File: src/com/cse3063/rag/orchestrator/Pipeline.java
import java.util.ArrayList;
import java.util.List;

/**
 * [GRASP - Pure Fabrication]
 * Responsible for the mechanics of running stages sequentially.
 * It is decoupled from the specific control flow of the Orchestrator.
 */
public class Pipeline {

    // The list of PipelineStage implementations to execute
    private final List<PipelineStage> stages = new ArrayList<>();

    public void addStage(PipelineStage stage) {
        this.stages.add(stage);
    }

    /**
     * Executes all stages added to the pipeline in order.
     * The Observer (TraceBus) logic should be integrated here or in the Orchestrator.
     */
    public RagContext execute(RagContext context) {
        for (PipelineStage stage : stages) {
            long startTime = System.currentTimeMillis();

            try {
                // Execute the stage logic
                context = stage.execute(context);

                // Publish trace event (Simulated logging)
                long duration = System.currentTimeMillis() - startTime;
                context.addMetadata(stage.getName() + "_time", duration);
                System.out.println(String.format("[TRACE] %s completed in %d ms.", stage.getName(), duration));

            } catch (Exception e) {
                // Publish error trace event
                context.addMetadata(stage.getName() + "_error", e.getMessage());
                System.err.println(String.format("[ERROR] Pipeline stopped at %s: %s", stage.getName(), e.getMessage()));
                break; // Stop pipeline execution on failure
            }
        }
        return context;
    }
}

File: python/rag/orchestrator/pipeline.py
"""Pipeline execution engine."""

from typing import List
import time
from .context import Context
from .pipeline_stage import PipelineStage
from ..tracer.trace_bus import TraceBus
from ..tracer.trace_event import TraceEvent


class Pipeline:
    """
    Pipeline for executing stages sequentially.
    [GRASP - Pure Fabrication]
    """
    
    def __init__(self):
        """Initialize empty pipeline."""
        self.stages: List[PipelineStage] = []
    
    def add_stage(self, stage: PipelineStage) -> None:
        """Add a stage to the pipeline."""
        self.stages.append(stage)
    
    def execute(self, context: Context, trace_bus: TraceBus) -> Context:
        """
        Execute all stages in order.
        
        Args:
            context: The context object to process
            trace_bus: Trace bus for logging events
            
        Returns:
            Updated context after all stages
        """
        for stage in self.stages:
            start_time = time.time()
            stage_name = stage.get_name()
            success = True
            output_summary = "Success"
            
            try:
                # Execute stage logic
                stage.execute(context)
                
            except Exception as e:
                # Handle failures gracefully
                success = False
                output_summary = f"Error: {str(e)}"
                print(f"[CRITICAL] Pipeline stopped at {stage_name}: {str(e)}")
                import traceback
                traceback.print_exc()
                
            finally:
                # Calculate duration
                duration = time.time() - start_time
                
                # Create trace event
                event = TraceEvent(
                    stage_name=stage_name,
                    input_summary=str(context),
                    output_summary=output_summary,
                    duration_ms=int(duration * 1000)
                )
                
                # Publish to trace bus
                if trace_bus:
                    trace_bus.trace(event)
                
                # Stop pipeline if stage failed
                if not success:
                    break
        
        return context
